File: lib/block.py
"""Data structure for a single block."""

import base64
import datetime
import json
# pylint: disable=unused-import
from typing import Any, Dict, List, Set, TYPE_CHECKING
# pylint: enable=unused-import

# pylint: disable=C0411,R0902
from .crypto import Hash512, Hasher, my_random_bytes
from .errors import (BlockDifficultyError, BlockNotFoundError,
                     BlockValidationError)
from .transaction import Transaction

if TYPE_CHECKING:
    # pylint: disable=unused-import
    from querylayer import QueryLayer
    # pylint: enable=unused-import


BLOCK_CURRENT_VERSION = 1
MAX_TRANSACTIONS_PER_BLOCK = 10
MAX_BLOCKSIZE = 4096
GENESIS_HASH = b"There can be only one"
NONCE_SIZE = 4


class Block:
    """Block class."""

    def __init__(self, index: int, prev_hash: Hash512) -> None:
        """Create a new Block object."""
        self.index = index
        self.version = BLOCK_CURRENT_VERSION

        now = datetime.datetime.now(datetime.timezone.utc)
        self.timestamp = int(now.timestamp())

        self.transactions = []  # type: List[Transaction]
        self.merkle_root = Hash512(b'')  # Root hash of Merkle Tree.

        self.prev_hash = prev_hash
        self.salt = my_random_bytes(NONCE_SIZE)

        self.nonce = my_random_bytes(NONCE_SIZE)
        self.hash = Hash512(b'')
        return

    @staticmethod
    def deserialise(data: str) -> 'Block':
        """Deserialise from bytes to a Block object."""
        d = json.loads(data)
        return Block.deserialise_dict(d)

    @staticmethod
    def deserialise_dict(d: Dict[str, Any]) -> 'Block':
        """Deserialise from dict to a Block object."""
        version = d['version']
        if version == 1:
            b = Block(
                index=d['index'],
                prev_hash=Hash512.deserialise(d['prevhash'])
            )
            b.version = version
            b.timestamp = d['ts']
            b.transactions = [Transaction.deserialise(x)
                              for x in d['transactions']]
            b.merkle_root = Hash512.deserialise(d['merkle'])
            b.nonce = base64.b64decode(d['nonce'])
            b.salt = base64.b64decode(d['salt'])
            b.hash = Hash512.deserialise(d['hash'])
            return b

        raise Exception("Unknown block version: {}".format(version))

    def serialise_dict(self) -> Dict[str, Any]:
        """Serialise this block to a dict."""
        return {
            'version': self.version,
            'index': self.index,
            'ts': self.timestamp,
            'transactions': self.get_serialised_transactions(),
            'merkle': self.merkle_root.serialise(),
            'prevhash': self.prev_hash.serialise(),
            'salt': base64.b64encode(self.salt).decode('utf-8'),
            'nonce': base64.b64encode(self.nonce).decode('utf-8'),
            'hash': self.hash.serialise()
        }

    def serialise(self) -> str:
        """Serialise this block to JSON str."""
        if self.version == 1:
            return json.dumps(self.serialise_dict())

        raise Exception("Unknown block version: {}".format(self.version))

    def get_serialised_transactions(self) -> List[str]:
        """Get serialised transaction data."""
        return [x.serialise() for x in self.transactions]

    def calculate_merkle_root(self) -> Hash512:
        """Calculate the merkle root hash."""
        hashes = [x.txid for x in self.transactions]

        while len(hashes) > 1:
            # If an odd number, repeat the last one.
            if len(hashes) % 2 != 0:
                hashes.append(hashes[-1])

            hashes_out = []
            for i in range(0, len(hashes), 2):
                h = Hasher([
                    hashes[i],
                    hashes[i+1]
                ]).get_hash()
                hashes_out.append(h)

            hashes = hashes_out
        return hashes[0]

    def calculate_hash(self) -> Hash512:
        """Calculate the hash."""
        return Hasher([
            str(self.version),
            str(self.index),
            str(self.timestamp),
            str(len(self.transactions)),
            self.merkle_root,
            self.prev_hash,
            self.salt,
            self.nonce
        ]).get_hash()

    def calculate_new_hash(self) -> Hash512:
        """Modify the nonce and recalculate the hash."""
        self.nonce = my_random_bytes(NONCE_SIZE)
        return self.calculate_hash()

    def verify_pow(self, pow_difficulty: int) -> None:
        """Verify that this block meets POW criteria."""
        hstr = self.hash.to_string()
        if hstr[:pow_difficulty] != '0' * pow_difficulty:
            raise BlockDifficultyError("Block difficulty too low")
        return

    def ensure_difficulty(self, pow_difficulty: int) -> None:
        """Keep recalculating the hash until difficulty level is met."""
        while True:
            try:
                self.verify_pow(pow_difficulty)
                break
            except BlockDifficultyError:
                self.hash = self.calculate_new_hash()
        return

    def validate(self, q: 'QueryLayer', pow_difficulty: int) -> bool:
        """Validate this block."""
        try:
            self.verify_pow(pow_difficulty)

            if self.index > 0:
                # Ensure transaction inputs are unique for the entire block.
                known_inputs = set()  # type: Set[Hash512]

                for trans in self.transactions:
                    trans.validate(q)

                    # Multiple transactions in the same block, referencing the
                    # same input, are not allowed!
                    trans.check_duplicates(known_inputs)

            assert self.version == 1, \
                "Unknown version {!r}".format(self.version)

            assert len(self.serialise()) <= MAX_BLOCKSIZE, \
                "Block exceeds maximum size of {} bytes".format(MAX_BLOCKSIZE)

            assert len(self.transactions) <= MAX_TRANSACTIONS_PER_BLOCK, \
                "Block contains too many transactions"

            d = datetime.datetime.fromtimestamp(self.timestamp)
            assert d > datetime.datetime(2018, 1, 1), \
                "Block timestamp is too old!"
            assert d < datetime.datetime.now(), \
                "Block timestamp is in the future!"

            assert self.index >= 0 and self.index <= q.get_num_blocks(), \
                "Invalid block index"

            if self.index == 0:
                # Compare genesis block against the blockchain.
                try:
                    genblock = q.get_block(0)
                    assert self.hash == genblock.hash, \
                        "Block hash mismatch"
                except BlockNotFoundError:
                    # Brand new blockchain?
                    assert q.get_num_blocks() == 0, "Block count mismatch"
                    assert self.hash == Hasher(GENESIS_HASH).get_hash(), \
                        "Block hash mismatch"
            else:
                prev_block = q.get_block(self.index - 1)
                assert self.prev_hash == prev_block.hash, \
                    "Previous block hash mismatch"

            assert self.merkle_root == self.calculate_merkle_root(), \
                "Block merkle root hash mismatch"

            assert self.hash == self.calculate_hash(), "Block hash mismatch"

        except (AssertionError, BlockDifficultyError) as exc:
            raise BlockValidationError(
                "Error validating block: {}".format(self.index)) from exc
        return True

File: rs/src/block.rs
/// Block data structure — direct port of lib/block.py.
use std::collections::HashSet;
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

use crate::crypto::{hash512, hash512_from_hex, random_bytes, Hash, Hasher};
use crate::errors::{BlockChainError, Result};
use crate::querylayer::QueryLayer;
use crate::transaction::Transaction;

pub const BLOCK_CURRENT_VERSION: u32 = 1;
pub const MAX_TRANSACTIONS_PER_BLOCK: usize = 10;
pub const MAX_BLOCKSIZE: usize = 4096;
pub const GENESIS_HASH: &[u8] = b"There can be only one";
const NONCE_SIZE: usize = 4;

#[derive(Clone, Debug)]
pub struct Block {
    pub index: u64,
    pub version: u32,
    pub timestamp: u64,
    pub transactions: Vec<Transaction>,
    pub merkle_root: Hash,
    pub prev_hash: Hash,
    pub salt: Vec<u8>,
    pub nonce: Vec<u8>,
    pub hash: Hash,
}

#[derive(Serialize, Deserialize)]
pub struct BlockJson {
    pub version: u32,
    pub index: u64,
    pub ts: u64,
    pub transactions: Vec<String>,
    pub merkle: String,
    pub prevhash: String,
    pub salt: String,
    pub nonce: String,
    pub hash: String,
}

impl Block {
    pub fn new(index: u64, prev_hash: Hash) -> Result<Self> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        Ok(Block {
            index,
            version: BLOCK_CURRENT_VERSION,
            timestamp: now,
            transactions: Vec::new(),
            merkle_root: hash512(b"")?,
            prev_hash,
            salt: random_bytes(NONCE_SIZE),
            nonce: random_bytes(NONCE_SIZE),
            hash: hash512(b"")?,
        })
    }

    pub fn deserialise(data: &str) -> Result<Self> {
        let d: BlockJson = serde_json::from_str(data)?;
        Self::deserialise_dict(&d)
    }

    pub fn deserialise_dict(d: &BlockJson) -> Result<Self> {
        if d.version != 1 {
            return Err(BlockChainError::BlockValidation(format!(
                "Unknown block version: {}",
                d.version
            )));
        }

        let transactions: Result<Vec<_>> =
            d.transactions.iter().map(|s| Transaction::deserialise(s)).collect();

        Ok(Block {
            version: d.version,
            index: d.index,
            timestamp: d.ts,
            transactions: transactions?,
            merkle_root: hash512_from_hex(&d.merkle)?,
            prev_hash: hash512_from_hex(&d.prevhash)?,
            nonce: base64::Engine::decode(
                &base64::engine::general_purpose::STANDARD,
                &d.nonce,
            )
            .map_err(|e| BlockChainError::BlockValidation(format!("Bad nonce: {e}")))?,
            salt: base64::Engine::decode(
                &base64::engine::general_purpose::STANDARD,
                &d.salt,
            )
            .map_err(|e| BlockChainError::BlockValidation(format!("Bad salt: {e}")))?,
            hash: hash512_from_hex(&d.hash)?,
        })
    }

    pub fn serialise_dict(&self) -> BlockJson {
        use base64::Engine;
        BlockJson {
            version: self.version,
            index: self.index,
            ts: self.timestamp,
            transactions: self
                .transactions
                .iter()
                .map(|t| t.serialise().unwrap_or_default())
                .collect(),
            merkle: self.merkle_root.serialise(),
            prevhash: self.prev_hash.serialise(),
            salt: base64::engine::general_purpose::STANDARD.encode(&self.salt),
            nonce: base64::engine::general_purpose::STANDARD.encode(&self.nonce),
            hash: self.hash.serialise(),
        }
    }

    pub fn serialise(&self) -> Result<String> {
        Ok(serde_json::to_string(&self.serialise_dict())?)
    }

    pub fn calculate_merkle_root(&self) -> Hash {
        let mut hashes: Vec<Hash> = self.transactions.iter().map(|t| t.txid.clone()).collect();

        while hashes.len() > 1 {
            // If odd, repeat the last one
            if hashes.len() % 2 != 0 {
                hashes.push(hashes.last().unwrap().clone());
            }

            let mut next = Vec::new();
            for pair in hashes.chunks(2) {
                let h =
                    Hasher::new_with_items(&[(&pair[0]).into(), (&pair[1]).into()]).get_hash();
                next.push(h);
            }
            hashes = next;
        }

        hashes.into_iter().next().unwrap()
    }

    pub fn calculate_hash(&self) -> Hash {
        let ver = self.version.to_string();
        let idx = self.index.to_string();
        let ts = self.timestamp.to_string();
        let txcount = self.transactions.len().to_string();

        Hasher::new_with_items(&[
            ver.as_str().into(),
            idx.as_str().into(),
            ts.as_str().into(),
            txcount.as_str().into(),
            (&self.merkle_root).into(),
            (&self.prev_hash).into(),
            self.salt.as_slice().into(),
            self.nonce.as_slice().into(),
        ])
        .get_hash()
    }

    pub fn calculate_new_hash(&mut self) -> Hash {
        self.nonce = random_bytes(NONCE_SIZE);
        self.calculate_hash()
    }

    pub fn verify_pow(&self, pow_difficulty: usize) -> Result<()> {
        let hstr = self.hash.to_hex();
        let prefix: String = std::iter::repeat('0').take(pow_difficulty).collect();
        if !hstr.starts_with(&prefix) {
            return Err(BlockChainError::BlockDifficulty);
        }
        Ok(())
    }

    pub fn ensure_difficulty(&mut self, pow_difficulty: usize) {
        loop {
            if self.verify_pow(pow_difficulty).is_ok() {
                break;
            }
            self.hash = self.calculate_new_hash();
        }
    }

    pub fn validate(&self, q: &mut QueryLayer, pow_difficulty: usize) -> Result<()> {
        self.verify_pow(pow_difficulty)?;

        if self.index > 0 {
            let mut known_inputs: HashSet<Hash> = HashSet::new();
            for trans in &self.transactions {
                trans.validate(q)?;
                trans.check_duplicates(&mut known_inputs)?;
            }
        }

        if self.version != 1 {
            return Err(BlockChainError::BlockValidation(format!(
                "Unknown version {}", self.version
            )));
        }

        let serialised = self.serialise()?;
        if serialised.len() > MAX_BLOCKSIZE {
            return Err(BlockChainError::BlockValidation(format!(
                "Block exceeds maximum size of {MAX_BLOCKSIZE} bytes"
            )));
        }

        if self.transactions.len() > MAX_TRANSACTIONS_PER_BLOCK {
            return Err(BlockChainError::BlockValidation(
                "Block contains too many transactions".into(),
            ));
        }

        // Timestamp checks
        // Must be after 2018-01-01 and before now
        let min_ts = 1_514_764_800u64; // 2018-01-01 UTC
        if self.timestamp < min_ts {
            return Err(BlockChainError::BlockValidation(
                "Block timestamp is too old".into(),
            ));
        }
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        if self.timestamp > now {
            return Err(BlockChainError::BlockValidation(
                "Block timestamp is in the future".into(),
            ));
        }

        let num_blocks = q.get_num_blocks();
        if self.index > num_blocks {
            return Err(BlockChainError::BlockValidation(
                "Invalid block index".into(),
            ));
        }

        if self.index == 0 {
            match q.get_block(0) {
                Ok(genblock) => {
                    if self.hash != genblock.hash {
                        return Err(BlockChainError::BlockValidation(
                            "Block hash mismatch".into(),
                        ));
                    }
                }
                Err(BlockChainError::BlockNotFound(_)) => {
                    if num_blocks != 0 {
                        return Err(BlockChainError::BlockValidation(
                            "Block count mismatch".into(),
                        ));
                    }
                }
                Err(e) => return Err(e),
            }
        } else if self.index <= num_blocks {
            let prev_block = q.get_block(self.index - 1)?;
            if self.prev_hash != prev_block.hash {
                return Err(BlockChainError::BlockValidation(
                    "Previous block hash mismatch".into(),
                ));
            }
        }

        if self.merkle_root != self.calculate_merkle_root() {
            return Err(BlockChainError::BlockValidation(
                "Block merkle root hash mismatch".into(),
            ));
        }

        if self.hash != self.calculate_hash() {
            return Err(BlockChainError::BlockValidation(
                "Block hash mismatch".into(),
            ));
        }

        Ok(())
    }
}
